Ensure `env_params` intake for calibration

leewujung opened the pull request:
## Overview

~This PR fixes a regression bug in v0.6.4 and ensures `env_params` are taken up correctly for EK60 echosounder.~

It turned out that the bug was not omission, but that the `env_params` used for EK60 cal in v0.6.4 was not passed in from the correctly sanitized ones from the parent class method. This is now in a PR with minimal scope #987.

Therefore, I am putting in work in this PR to address #965 more comprehensively.


## Details
- Overhaul functions in `env_params.py` to address the core of #965
- Remove redundancy in passing in extra `ed_beam_group` in `_cal_power_samples` and `_cal_complex_samples`
- Create integration tests for the `get_env_params_*` functions in `env_params_integration.py`
  - These verify that user-provided params are taken in to the calibration objects and stored in the final Sv output correctly
- Create unit tests for the `get_env_params_*` functions in `env_params.py`

leewujung commented:
@emiliom : I think this is ready for a review when you are available. I will add that one integration test tomorrow morning.

leewujung commented:
@emiliom : This one is ready for a thorough review.

emiliom commented on echopype/calibrate/calibrate_ek.py:
```diff
@@ -130,22 +130,27 @@ class CalibrateEK60(CalibrateEK):
 
-        # Set sonar_type
+        # Set sonar_type and waveform/encode mode
         self.sonar_type = "EK60"
```
Small comment: we use `sonar_model` in the echodata class and elsewhere. If it's not too disruptive, I'd suggest calling it that here, too, to maintain consistency across echopype and reduce cognitive overload.

leewujung replied:
Hmm, I used `sonar_type` because here I use "EK60" to stand for the older style of data, and "EK80" to stand for the newer style of data. Perhaps I can use "Ex60" (though this does not include ES70) and "Ex80" if we use `sonar_model`. Thoughts?

emiliom replied:
Hmm, I don't know what "Ex60" and "Ex80" are, so I can't comment on that.

But this may be related to the broader relationship between Simrad-Kongsberg `sonar_model` entries (eg, "EK60", "ES70") and the broader sonar "type" ("EK60" vs "EK80") that determines the parser, set_group_, and many other processing decisions. I've created #995 just now to describe this relationship and a need for describing it centrally in an easily reusable way.

But back to the question at hand, *here.* No need to belabor my minor naming point. If you think `sonar_type` is more appropriate, stick with it. Or if you think this really is related to #995, stick with `sonar_type` and let's discuss the broader I raised in #995 issue after 0.7.0.

leewujung replied:
"Ex60" and "Ex80" are what you called "class" in #995. I think we can keep it as is here, and address those code consolidation changes later in a more systematic way.

emiliom replied:
sounds good

emiliom commented on echopype/calibrate/calibrate_ek.py:
```diff
@@ -130,22 +130,27 @@ class CalibrateEK60(CalibrateEK):
-        self.compute_echo_range()
-
         # Get the right ed_group for CW power samples
         self.ed_group = retrieve_correct_beam_group(
```
Consider renaming `ed_group` to, say, `ed_beam_group`? Just for clarity.

leewujung replied:
Sure, this is something inherited, but I agree to change it for clarify.

leewujung replied:
Done! In the process I discovered a redundancy in input argument for `_cal_complex_samples` and `_cal_power_samples` and removed them as well.

emiliom commented on echopype/calibrate/env_params.py:
```diff
@@ -59,199 +68,289 @@ def harmonize_env_param_time(
-def get_env_params_AZFP(echodata: EchoData, user_env_dict: Optional[dict] = None):
-    """Get env params using user inputs or values from data file.
+def sanitize_user_env_dict(
+    user_dict: Dict[str, Union[int, float, list, xr.DataArray]],
+    channel: Union[List, xr.DataArray],
```
Any reason to use `list` in one case and `typing` `List` in the other?

emiliom replied:
Actually, given how frequently you're using the combination of a list and a dataarray, I suggest you define a global variable for that union, then reuse that. It'll make for more compact and consistent code. eg:

```python
sequence = Union[List, xr.DataArray]

# ...

user_dict: Dict[str, Union[int, float, sequence]],
channel: sequence,
```

leewujung replied:
But if we do this, the type hint users get would also be `sequence`, which may not be clear to users?

leewujung replied:
Apparently, I have not finished the test cases I planned to do!! 🤯 Good that you caught this!

emiliom replied:
That's a very good point. The strategy I suggested may not be a net benefit for user-facing functions, for the reason you bring up. But it may be a net benefit for internal functions, where code consistency is a higher priority.

So, scratch that. My only remaining comment here is the original one, which is very minor:

> Any reason to use `list` in one case and `typing` `List` in the other?

leewujung replied:
Oh i forgot the `list` part. I'll change that to `List` in the function typing, for consistency.

emiliom commented on echopype/calibrate/env_params.py:
```diff
@@ -59,199 +68,289 @@ def harmonize_env_param_time(
+    user_dict : dict
+        A dictionary containing user input calibration parameters
+        as {parameter name: parameter value}.
+        Parameter value has to be a scalar (int or float) or an ``xr.DataArray``.
```
```suggestion
        Parameter value has to be a scalar (int or float), a list or an ``xr.DataArray``.
```

emiliom commented on echopype/calibrate/env_params.py:
```diff
@@ -59,199 +68,289 @@ def harmonize_env_param_time(
+            # Param "sound_absorption" has to be an xr.DataArray or a list because it is freq-dep
+            if p_name == "sound_absorption" and not (p_val, (xr.DataArray, list)):
+                raise ValueError(
+                    "The 'sound_absorption' parameter has to be an xr.DataArray, "
```
So, does it have to be an xr.DataArray or either it or a list, per the comment above?

leewujung replied:
Yes, I forgot to change this.
```suggestion
                    "The 'sound_absorption' parameter has to be a list or an xr.DataArray, "
```

emiliom commented on echopype/calibrate/env_params.py:
```diff
@@ -59,199 +68,289 @@ def harmonize_env_param_time(
+    for p_name, p_val in user_dict.items():
+        if p_name in out_dict:
+            # Param "sound_absorption" has to be an xr.DataArray or a list because it is freq-dep
+            if p_name == "sound_absorption" and not (p_val, (xr.DataArray, list)):
```
Did you mean this?
```suggestion
            if p_name == "sound_absorption" and not isinstance(p_val, (xr.DataArray, list)):
```

leewujung replied:
Oh geez, yes! I don't know how I missed that! And I wonder why the test passes through!

leewujung replied:
Apparently, I have not finished the test cases I planned to do!! 🤯 Good that you caught this!

emiliom commented on echopype/calibrate/env_params.py:
```diff
@@ -59,199 +68,289 @@ def harmonize_env_param_time(
+        "FG",
+    ]:
+        raise ValueError("'formula_absorption' has to be None, 'FG' or 'AM' for EK echosounders.")
+    if (
+        out_dict["formula_sound_speed"] is not None
+        and out_dict["formula_sound_speed"] != "Mackenzie"
+    ):
```
I _think_ this works with `None`, too
```suggestion
    if out_dict["formula_sound_speed"] not in (None,  "Mackenzie"):
```

leewujung replied:
Oh you are right, it does work with `None`! Great, this makes it so much more succint.

leewujung replied:
I made the same change to the `formula_sound_absorption` if statement above too

emiliom commented on echopype/calibrate/env_params.py:
```diff
@@ -59,199 +68,289 @@ def harmonize_env_param_time(
+            out_dict.pop("formula_sound_speed")
+        else:
+            # default to Mackenzie sound speed formula if not in user dict
+            if out_dict["formula_sound_speed"] is None:
+                out_dict["formula_sound_speed"] = "Mackenzie"
```
see my suggestion and comment above
```suggestion
            if out_dict["formula_sound_speed"] in (None, "Mackenzie"):
```

leewujung replied:
I added this but this was actually to set the `formula_sound_speed` to Mackenzie when it does not exist, not an if statement like anove. I've reverted it back.

emiliom reviewed (commented):
I wasn't thorough, but I think it looks good. I made a bunch of comments, most of them stylistic. 

There's just one where I think you're missing something important, an `inistance()`. Take a look.

leewujung commented on echopype/calibrate/env_params.py:
```diff
@@ -59,199 +68,289 @@ def harmonize_env_param_time(
+    out_dict = sanitize_user_env_dict(user_dict=user_dict, channel=beam["channel"])
+
+    # Check absorption and sound speed formula
+    if out_dict["formula_absorption"] is not None and not out_dict["formula_absorption"] in [
+        "AM",
+        "FG",
+    ]:
```
```suggestion
    if out_dict["formula_absorption"] not in [None, "AM", "FG"]:
```

leewujung replied:
I've added this separately.

leewujung commented:
@emiliom : Thanks for reviewing this! I've made the following changes:
- committed all but one of the suggested changes (that one was incorrect even though the pattern looks similar, so I reverted)
- added missed test cases for `test_sanitize_user_env_dict`
- remove redundancy in passing in `ed_beam_group` in `_cal_power_samples` and `_cal_complex_samples`

I had questions about two of your comments so left them open.
